File: ur_typical90/typical90_af.py
from itertools import permutations


def main():
    bad = [[0] * N for _ in range(N)]
    for x, y in XY:
        bad[x - 1][y - 1] = 1
        bad[y - 1][x - 1] = 1

    ans = 10001
    for lst in permutations(range(N)):
        time = 0
        for i, li in enumerate(lst):
            if i > 0 and bad[lst[i-1]][li]:
                break
            time += A[li][i]
        else:
            ans = min(ans, time)
    return print(ans if ans < 10001 else -1)


if __name__ == '__main__':
    N = int(input())
    A = [list(map(int, input().split())) for _ in range(N)]
    M = int(input())
    XY = [list(map(int, input().split())) for _ in range(M)]

    main()

File: ur_typical90/typical90_af.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using vecll = vector<ll>;
using matll = vector<vector<ll>>;

#define rep(i, n) for (ll i = 0; i < ll(n); i++)

int main() {
    ll n, m;
    cin >> n;
    matll a(n, vecll(n, 0));
    rep(i, n) rep(j, n) cin >> a[i][j];
    cin >> m;
    matll bad(n, vecll(n, 0));
    rep(i, m) {
        ll x, y;
        cin >> x >> y;
        bad[x - 1][y - 1]++, bad[y - 1][x - 1]++;
    }

    ll ans = 10001;
    vecll perm(n);
    rep(i, n) perm[i] = i;

    do {
        ll time = 0;
        rep(i, n) time += a[perm[i]][i];
        rep(i, n - 1) if (bad[perm[i]][perm[i + 1]]) time = 10001;
        ans = min(ans, time);
    } while (next_permutation(perm.begin(), perm.end()));

    if (ans >= 10001) ans = -1;
    cout << ans << endl;
    return 0;
}
